fuel-txpool/src/service.rs

```diff
@@ -1,143 +1,137 @@
+        let join_handle = join.take();
+        if let Some(join_handle) = join_handle {
+            let _ = self.sender.send(TxPoolMpsc::Stop).await;
+            let receiver = self.receiver.clone();
+            Some(tokio::spawn(async move {
+                let ret = join_handle.await;
+                *receiver.lock().await = ret.ok();
+                ()
+            }))
+        } else {
+            None
         }
-        res
     }
 
-    async fn find_one(&self, hash: &TxId) -> Option<TxInfo> {
-        self.txpool.read().await.txs().get(hash).cloned()
+    pub fn subscribe_ch(&self) -> broadcast::Receiver<TxStatusBroadcast> {
+        self.broadcast.subscribe()
     }
 
-    /// find all dependent tx and return them with requsted dependencies in one list sorted by Price.
-    async fn find_dependent(&self, hashes: &[TxId]) -> Vec<ArcTx> {
-        let mut seen = HashMap::new();
-        {
-            let pool = self.txpool.read().await;
-            for hash in hashes {
-                if let Some(tx) = pool.txs().get(hash) {
-                    pool.dependency()
-                        .find_dependent(tx.tx().clone(), &mut seen, pool.txs());
-                }
-            }
-        }
-        let mut list: Vec<ArcTx> = seen.into_iter().map(|(_, tx)| tx).collect();
-        // sort from high to low price
-        list.sort_by_key(|tx| Reverse(tx.gas_price()));
+    pub fn sender(&self) -> &mpsc::Sender<TxPoolMpsc> {
+        &self.sender
+    }
 
-        list
+    pub async fn insert(
+        &self,
+        txs: Vec<Arc<Transaction>>,
+    ) -> oneshot::Receiver<Vec<anyhow::Result<Vec<Arc<Transaction>>>>> {
+        let (response, receiver) = oneshot::channel();
+        let _ = self.sender.send(TxPoolMpsc::Insert { txs, response }).await;
+        receiver
     }
 
-    /// Iterete over `hashes` and return all hashes that we dont have.
-    async fn filter_by_negative(&self, tx_ids: &[TxId]) -> Vec<TxId> {
-        let mut res = Vec::new();
-        let pool = self.txpool.read().await;
-        for tx_id in tx_ids {
-            if pool.txs().get(tx_id).is_none() {
-                res.push(*tx_id)
-            }
-        }
-        res
+    pub async fn find(&self, ids: Vec<TxId>) -> oneshot::Receiver<Vec<Option<TxInfo>>> {
+        let (response, receiver) = oneshot::channel();
+        let _ = self.sender.send(TxPoolMpsc::Find { ids, response }).await;
+        receiver
     }
 
-    /// Return all sorted transactions that are includable in next block.
-    /// This is going to be heavy operation, use it with only when needed.
-    async fn includable(&self) -> Vec<ArcTx> {
-        let pool = self.txpool.read().await;
-        pool.sorted_includable()
+    pub async fn find_one(&self, id: TxId) -> oneshot::Receiver<Option<TxInfo>> {
+        let (response, receiver) = oneshot::channel();
+        let _ = self.sender.send(TxPoolMpsc::FindOne { id, response }).await;
+        receiver
```

Comment:
We should await the oneshot here to clean up the interface (and not require consumers to use `.await.await`)

```suggestion
        receiver.await
```

Reply:
https://github.com/FuelLabs/fuel-core/pull/407#discussion_r900116952